### PR TITLE
fix(core): propagation with nullable 1:1 relation

### LINKED ISSUE #3850: Nullable 1-to-1 entity propagation not working
**Describe the bug**
When assigning an entity to another in a 1-to-1 relationship (either way round) with an auto-generated primary key, the foreign key on the owner side does not get set at all when flushed.

After inspecting the queries generated, the foreign key is always being set to `NULL` after the new entity is assigned and flushed. Is this because the ID doesn't exist yet on the new entity (as ID is auto-generated)? In fact the 2 queries needed are:
- Insert new entity
- Assign new entity ID to foreign key
but it looks like the update is query is attempted first (so the ID does not exist yet), then the new entity is inserted. This is probably the issue.

I'm just trying to upgrade to v5, and this definitely used to work in v4 as the tests used to pass, but now in this fairly simple case, entities are not linked at all. Maybe I'm doing something wrong. I did check the docs but may have missed something.

**To Reproduce**
1. See reproduction: https://github.com/alisd23/mikro-orm-issues/tree/1-to-1-propagation-issue
2. Run `yarn` to install (docker-compose also needed)
3. Run `yarn test:watch`

**Expected behavior**
Expect the entities to be assigned in the database via the foreign key.

**Additional context**

**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| typescript | 4.9 |
| mikro-orm | 5.6.0 |
| your-driver | mysql |

### PR DESCRIPTION
**Pull request by alisd23**
Fixes #3850

I'm not very familar with the code, so I'm not confident with the solution, but I think this is what is needed.

Tests added and related tests for #3614 also still pass

### CODE REVIEW

File: packages/core/src/unit-of-work/UnitOfWork.ts [hunk 1]
```diff
@@ -907,7 +907,13 @@ export class UnitOfWork {
     const props = changeSet.meta.uniqueProps;
 
     for (const prop of props) {
-      if (prop.name in changeSet.payload && inserts.find(c => Utils.equals(c.payload[prop.name], changeSet.originalEntity![prop.name as string]))) {
+      const insert = inserts.find(c => Utils.equals(c.payload[prop.name], changeSet.originalEntity![prop.name as string]));
+
+      if (
+        prop.name in changeSet.payload &&
+        insert &&
+        insert.payload[prop.name]
```
Comment (B4nan) [inline packages/core/src/unit-of-work/UnitOfWork.ts:915]:
```suggestion
        insert?.payload[prop.name]
```

**Review by B4nan — APPROVED** [review]
Amazing, thank you! I just briefly checked the repros but didn't have time to actually look into it just yet. The change looks fine to me, so as long as the test suite is passing, we are good to merge.

**Comment by alisd23** [comment]
No worries - just give me a few hours so I can add more test cases as I'm not quite sure this is the right solution.

Let me know if you've had time to check whether you think it's right/wrong 👍

**Comment by alisd23** [comment]
Added a few more test cases, so now we cover:
- Adding a new entity to the relation (either side)
- Swapping an entity on the relation (either side)
- Unlinking an entity - but not deleting either entity (either side)
- Deleting an entity - (either side)

All seems to work as expected - hopefully nothing broken downstream

File: packages/core/src/unit-of-work/UnitOfWork.ts (hunk 1 above)
Comment (alisd23) [reply packages/core/src/unit-of-work/UnitOfWork.ts:915]: I can add this if you like, but shouldn't be needed due to the previous `&& insert` check on line 914.

Could instead replace as `insert.payload?[prop.name]` if you want to be extra careful about the payload being null/undefined.

Comment (alisd23) [reply packages/core/src/unit-of-work/UnitOfWork.ts:915]: I've tried to have more of a think, and I believe this may be working by chance, though the tests work - I don't think the logic actually directly relates to the issue.

The original solution only breaks on my first test case: creating a new inverse entity with a auto-generated primary key.

I think the actual issue is specifically checking for if a primary key exists on the new entity (the insert), and if it doesn't then we *need* to insert that entity first (i.e. no early update)

Comment (alisd23) [reply packages/core/src/unit-of-work/UnitOfWork.ts]: Added a new but similar solution (plus extra tests)
- Let me know if you would like the new test file renamed or moved elsewhere. I couldn't think how else to add it in the current test structure
- Also let me know what you think of the newest solution, as I'm not clear on all the cases for `UPDATE_EARLY` changes when they were added initially

Comment (B4nan) [reply packages/core/src/unit-of-work/UnitOfWork.ts]: > I can add this if you like, but shouldn't be needed due to the previous && insert check on line 914.

FYI I proposed to replace those two lines with just one, not to use optional chaining for no reason :] 

> Also let me know what you think of the newest solution, as I'm not clear on all the cases for UPDATE_EARLY changes when they were added initially

Whatever makes the test suite green is fine with me. Early updates are there to handle edge cases mainly with unique column updates, I have been literally just fixing things and had to introduce it, so its not a thorough concept or something :]

File: tests/issues/GH3850-required-relation.test.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+  @OneToOne(() => User, user => user.project, { nullable: false })
+  public owner?: User;
+
+}
+
+let orm: MikroORM;
+
+beforeAll(async () => {
+  orm = await MikroORM.init({
```
Comment (B4nan) [inline tests/issues/GH3850-required-relation.test.ts:38]: this test hangs because its not calling `orm.close`

also please don't use `public` modified, its implicit

**Comment by alisd23** [comment]
@B4nan All comments are addressed now. I'm happier with the latest solution as I think it's more specific and makes more sense in my head. There's a hell of a lot of tests which is good, so hopefully it's not broken anything related to 1-M or M-M relations.

Feel free to squash all commits into the first when merging - thanks!

File: tests/issues/GH3850.test.ts [hunk 3]
```diff
@@ -47,6 +47,8 @@ beforeEach(async () => {
   await orm.schema.refreshDatabase();
 });
 
+afterAll(() => orm.close(true));
```
Comment (B4nan) [inline tests/issues/GH3850.test.ts:50]: this test already calls `close` on line 74/76, calling it twice will probably fail

Comment (alisd23) [reply tests/issues/GH3850.test.ts:50]: My bad - fixed

**Comment by alisd23** [comment]
All looks good now on CI 👍 

Any idea when we might be able to release the fix? This is blocking the upgrade on my project unfortunately

**Comment by B4nan** [comment]
every commit to master is published to NPM as dev version, so you can try it in few minutes

i will probably do a patch release on sunday
